python/lsst/resources/_resourceHandles/_s3ResourceHandle.py

```diff
@@ -0,0 +1,260 @@
+                # indicate that the object is in closing status
+                self._closed = CloseStatus.CLOSING
+                self.flush()
+                with time_this(self._log, msg="Finalize multipart upload to %s", args=(self,)):
+                    self._client.complete_multipart_upload(
+                        Bucket=self._multiPartUpload["Bucket"],
+                        Key=self._multiPartUpload["Key"],
+                        UploadId=self._multiPartUpload["UploadId"],
+                        MultipartUpload={"Parts": self._parts},
+                    )
+            else:
+                # Put the complete object at once
+                with time_this(self._log, msg="Write to %s", args=(self,)):
+                    self._client.put_object(Bucket=self._bucket, Key=self._key, Body=self._buffer.getvalue())
+        self._closed = CloseStatus.CLOSED
+
+    @property
+    def closed(self) -> bool:
+        return self._closed == CloseStatus.CLOSED
+
+    def fileno(self) -> int:
+        raise OSError("S3 object does not have a file number")
+
+    def flush(self) -> None:
+        # If the object is closed, not writeable, or rw flush should be skipped
+        # rw mode skips flush because the whole bytestream must be kept in
+        # the buffer for seeking reasons.
+        if self.closed or not self.writable() or "+" in self._mode:
+            return
+        # Disallow writes to seek to a position prior to the previous flush
+        # this allows multipart uploads to upload content as the stream is
+        # written to.
+        if (
+            (self.tell() - (self._last_flush_position or 0)) < 5242880
+            and not self._closed == CloseStatus.CLOSING
+            and not self._warned
+        ):
+            warnings.warn("S3 does not support flushing objects less than 5Mb, skipping")
+            self._warned = True
+            return
+        # nothing to write, don't create an empty upload
+        if self.tell() == 0:
+            return
+        with time_this(
+            self._log,
+            msg="Upload multipart %d to %s",
+            args=(
+                self._partNo,
+                self,
+            ),
+        ):
+            response = self._client.upload_part(
+                Body=self._buffer.getvalue(),
+                Bucket=self._bucket,
+                Key=self._key,
+                UploadId=self._multiPartUpload["UploadId"],
+                PartNumber=self._partNo,
+            )
+        self._parts.append({"PartNumber": self._partNo, "ETag": response["ETag"]})
+        self._partNo += 1
+        self._last_flush_position = self._buffer.tell() + (self._last_flush_position or 0)
+        self._buffer = BytesIO()
+
+    @property
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return self._readable
+
+    def readline(self, size: int = -1) -> bytes:
+        raise OSError("S3 Does not support line by line reads")
```

Comment:
The complexity of this with write, flush, and seek was large enough I didn't feel it was worth it vs the user just reading in the whole file and splitting how they chose, but if you wish I can go and have a stab at it

Reply:
At this point I think it's more important to get this work merged. Can you file a ticket for improving readline (and maybe add it to the message here so that people know where to comment if they hit that error).